File: sdk/go/posthog.go
package sdk

import (
	"bytes"
	"context"
	"encoding/json"
	"log"
	"net/http"
	"time"

	"github.com/open-feature/go-sdk/openfeature"
)

// PostHogHook implements the openfeature.Hook interface to send evaluation events to PostHog.
type PostHogHook struct {
	posthogAPIKey string
	endpoint      string
	client        *http.Client
}

// NewPostHogHook creates a new PostHogHook with the provided API key.
// By default it sends to https://us.i.posthog.com/capture/
func NewPostHogHook(posthogAPIKey string, endpoint string) *PostHogHook {
	if endpoint == "" {
		endpoint = "https://us.i.posthog.com/capture/"
	}
	return &PostHogHook{
		posthogAPIKey: posthogAPIKey,
		endpoint:      endpoint,
		client:        &http.Client{Timeout: 5 * time.Second},
	}
}

func (h *PostHogHook) Before(ctx context.Context, hookContext openfeature.HookContext, hookHints openfeature.HookHints) (*openfeature.EvaluationContext, error) {
	return nil, nil
}

func (h *PostHogHook) After(ctx context.Context, hookContext openfeature.HookContext, flagEvaluationDetails openfeature.InterfaceResolutionDetail, hookHints openfeature.HookHints) error {
	distinctID := hookContext.EvaluationContext().TargetingKey()
	if distinctID == "" {
		distinctID = "anonymous-server-user"
	}

	payload := map[string]interface{}{
		"api_key": h.posthogAPIKey,
		"event":   "$feature_flag_called",
		"properties": map[string]interface{}{
			"distinct_id":            distinctID,
			"$feature_flag":          hookContext.FlagKey(),
			"$feature_flag_response": flagEvaluationDetails.Value,
			"timestamp":              time.Now().UTC().Format(time.RFC3339),
		},
	}

	go h.sendEvent(payload)
	return nil
}

func (h *PostHogHook) Error(ctx context.Context, hookContext openfeature.HookContext, err error, hookHints openfeature.HookHints) {
	distinctID := hookContext.EvaluationContext().TargetingKey()
	if distinctID == "" {
		distinctID = "anonymous-server-user"
	}

	payload := map[string]interface{}{
		"api_key": h.posthogAPIKey,
		"event":   "Feature Flag Error",
		"properties": map[string]interface{}{
			"distinct_id":   distinctID,
			"$feature_flag": hookContext.FlagKey(),
			"error":         err.Error(),
			"timestamp":     time.Now().UTC().Format(time.RFC3339),
		},
	}

	go h.sendEvent(payload)
}

func (h *PostHogHook) Finally(ctx context.Context, hookContext openfeature.HookContext, hookHints openfeature.HookHints) {
	// No-op
}

func (h *PostHogHook) sendEvent(payload map[string]interface{}) {
	body, err := json.Marshal(payload)
	if err != nil {
		log.Printf("[flagmanagment-posthog] failed to marshal event: %v", err)
		return
	}

	req, err := http.NewRequest("POST", h.endpoint, bytes.NewBuffer(body))
	if err != nil {
		return
	}

	req.Header.Set("Content-Type", "application/json")
	resp, err := h.client.Do(req)
	if err != nil {
		log.Printf("[flagmanagment-posthog] failed to send event: %v", err)
		return
	}
	defer resp.Body.Close()
}

File: sdk/python/flagmanagment/posthog.py
import logging
import requests
import threading
from datetime import datetime, timezone
from openfeature.hook import Hook
from openfeature.evaluation_context import EvaluationContext
from openfeature.hook_context import HookContext
from openfeature.hook_hints import HookHints
from openfeature.flag_evaluation import FlagEvaluationDetails

logger = logging.getLogger("flagmanagment-posthog")

class PostHogHook(Hook):
    def __init__(self, posthog_api_key: str, endpoint: str = "https://us.i.posthog.com/capture/"):
        self.posthog_api_key = posthog_api_key
        self.endpoint = endpoint

    def after(self, hook_context: HookContext, details: FlagEvaluationDetails, hints: HookHints):
        context: EvaluationContext = hook_context.evaluation_context
        distinct_id = context.targeting_key if context and context.targeting_key else "anonymous-server-user"

        payload = {
            "api_key": self.posthog_api_key,
            "event": "$feature_flag_called",
            "properties": {
                "distinct_id": distinct_id,
                "$feature_flag": hook_context.flag_key,
                "$feature_flag_response": details.value,
                "timestamp": datetime.now(timezone.utc).isoformat()
            }
        }
        self._send_event_async(payload)

    def error(self, hook_context: HookContext, exception: Exception, hints: HookHints):
        context: EvaluationContext = hook_context.evaluation_context
        distinct_id = context.targeting_key if context and context.targeting_key else "anonymous-server-user"

        payload = {
            "api_key": self.posthog_api_key,
            "event": "Feature Flag Error",
            "properties": {
                "distinct_id": distinct_id,
                "$feature_flag": hook_context.flag_key,
                "error": str(exception),
                "timestamp": datetime.now(timezone.utc).isoformat()
            }
        }
        self._send_event_async(payload)

    def _send_event_async(self, payload: dict):
        def _send():
            try:
                requests.post(self.endpoint, json=payload, timeout=5)
            except Exception as e:
                logger.warning(f"Failed to send PostHog event: {e}")
        
        threading.Thread(target=_send, daemon=True).start()
